src/main/java/com/google/devtools/build/lib/remote/ExecuteRetrier.java

```diff
@@ -0,0 +1,127 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.devtools.build.lib.remote;
+
+import com.google.common.util.concurrent.ListeningScheduledExecutorService;
+import com.google.protobuf.Any;
+import com.google.protobuf.InvalidProtocolBufferException;
+import com.google.protobuf.util.Durations;
+import com.google.rpc.DebugInfo;
+import com.google.rpc.Help;
+import com.google.rpc.LocalizedMessage;
+import com.google.rpc.RetryInfo;
+import com.google.rpc.RequestInfo;
+import com.google.rpc.ResourceInfo;
+import com.google.rpc.PreconditionFailure;
+import com.google.rpc.PreconditionFailure.Violation;
+import io.grpc.Status.Code;
+import io.grpc.protobuf.StatusProto;
+
+class ExecuteRetrier extends RemoteRetrier {
+
+  private static final String VIOLATION_TYPE_MISSING = "MISSING";
+
+  private static class RetryInfoBackoff implements Backoff {
+    private final int maxRetryAttempts;
+    int retries = 0;
+
+    RetryInfoBackoff(int maxRetryAttempts) {
+      this.maxRetryAttempts = maxRetryAttempts;
+    }
+
+    @Override
+    public long nextDelayMillis(Exception e) {
+      if (retries >= maxRetryAttempts) {
+        return -1;
+      }
+      RetryInfo retryInfo = getRetryInfo(e);
+      retries++;
+      return Durations.toMillis(retryInfo.getRetryDelay());
+    }
+
+    RetryInfo getRetryInfo(Exception e) {
+      RetryInfo retryInfo = RetryInfo.getDefaultInstance();
+      com.google.rpc.Status status = StatusProto.fromThrowable(e);
+      if (status != null) {
+        for (Any detail : status.getDetailsList()) {
+          if (detail.is(RetryInfo.class)) {
+            try {
+              retryInfo = detail.unpack(RetryInfo.class);
+            } catch (InvalidProtocolBufferException protoEx) {
+              // really shouldn't happen, ignore
+            }
+          }
+        }
+      }
+      return retryInfo;
+    }
+
+    @Override
+    public int getRetryAttempts() {
+      return retries;
+    }
+  }
+
+  ExecuteRetrier(
+      int maxRetryAttempts,
+      ListeningScheduledExecutorService retryService,
+      CircuitBreaker circuitBreaker) {
+    super(() -> maxRetryAttempts > 0 ? new RetryInfoBackoff(maxRetryAttempts) : RETRIES_DISABLED, ExecuteRetrier::test, retryService, circuitBreaker);
+  }
+
+  private static boolean test(Exception e) {
+    if (BulkTransferException.isOnlyCausedByCacheNotFoundException(e)) {
+      return true;
+    }
+    com.google.rpc.Status status = StatusProto.fromThrowable(e);
+    if (status == null || status.getDetailsCount() == 0) {
+      return false;
+    }
+    boolean fullyRetriable = false;
+    boolean failedPrecondition = status.getCode() == Code.FAILED_PRECONDITION.value();
+    for (Any detail : status.getDetailsList()) {
+      if (detail.is(RetryInfo.class)) {
+        // server says we can retry, regardless of other details
+        fullyRetriable = true;
```

Comment:
`break` here?

Reply:
Early return here - nothing after it is effective, and coming up with whether the precondition failure is effective is meaningless. Thanks!